Sources/MongoSwift/BSON/BsonDecoder.swift

```diff
@@ -154,14 +178,22 @@ extension _BsonDecoder {
         return typed
     }
 
-    fileprivate func unboxPrimitive<T: Primitive>(_ value: BsonValue?, as type: T.Type) throws -> T? {
+    fileprivate func unboxNumber<T: CodableNumber>(_ value: BsonValue?, as type: T.Type) throws -> T? {
         guard let unwrapped = value else {
             throw DecodingError._typeMismatch(at: self.codingPath, expectation: type, reality: value)
         }
-        return try T(from: unwrapped)
+
+        guard let primitive = T(from: unwrapped) else {
+            throw DecodingError._numberMismatch(at: self.codingPath, expectation: type, reality: value)
+        }
+        return primitive
     }
 
     fileprivate func unbox<T: Decodable>(_ value: BsonValue?, as type: T.Type) throws -> T? {
+        // if the data is already stored as the correct type in the document, then we can short-circuit
```

Comment:
this check is to support the case we discussed in SWIFT-86 where we are given extended JSON input, and so the value is already decoded to the right type and doesn't need to be initialized from the decoder.